test/core/event_engine/test_suite/tests/client_test.cc

```diff
@@ -293,5 +302,110 @@ TEST_F(EventEngineClientTest, MultipleIPv6ConnectionsToOneOracleListenerTest) {
   server_endpoint.reset();
 }
 
+// Create a connection using the test EventEngine to a listener created by the
+// test EventEngine and exchange bi-di data over the connection. Each endpoint
+// gets resetted as soon as the write is done. This test checks that EventEngine
+// implementations handle lifetimes around endpoints correctly.
+
+TEST_F(EventEngineClientTest, WriteEventCallbackTest) {
+  grpc_core::ExecCtx ctx;
+  std::shared_ptr<EventEngine> test_ee(this->NewEventEngine());
+  auto memory_quota = std::make_unique<grpc_core::MemoryQuota>("bar");
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  auto resolved_addr = URIToResolvedAddress(target_addr);
+  CHECK_OK(resolved_addr);
+  std::unique_ptr<EventEngine::Endpoint> client_endpoint;
+  std::unique_ptr<EventEngine::Endpoint> server_endpoint;
+  grpc_core::Notification client_signal;
+  grpc_core::Notification server_signal;
+
+  Listener::AcceptCallback accept_cb =
+      [&server_endpoint, &server_signal](
+          std::unique_ptr<Endpoint> ep,
+          grpc_core::MemoryAllocator /*memory_allocator*/) {
+        server_endpoint = std::move(ep);
+        server_signal.Notify();
+      };
+
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  ChannelArgsEndpointConfig config(args);
+  auto listener = *test_ee->CreateListener(
+      std::move(accept_cb),
+      [](absl::Status status) {
+        ASSERT_TRUE(status.ok()) << status.ToString();
+      },
+      config, std::make_unique<grpc_core::MemoryQuota>("foo"));
+
+  ASSERT_TRUE(listener->Bind(*resolved_addr).ok());
+  ASSERT_TRUE(listener->Start().ok());
+
+  test_ee->Connect(
+      [&client_endpoint,
+       &client_signal](absl::StatusOr<std::unique_ptr<Endpoint>> endpoint) {
+        ASSERT_TRUE(endpoint.ok());
+        client_endpoint = std::move(*endpoint);
+        client_signal.Notify();
+      },
+      *resolved_addr, config, memory_quota->CreateMemoryAllocator("conn-1"),
+      24h);
+
+  client_signal.WaitForNotification();
+  server_signal.WaitForNotification();
+  ASSERT_NE(client_endpoint.get(), nullptr);
+  ASSERT_NE(server_endpoint.get(), nullptr);
+
+  // Start writes with WriteEventCallbacks from the client endpoint and server
+  // endpoint and reset both endpoints immediately. It doesn't matter if the
+  // callbacks don't get invoked as long as there is no use-after-free behavior.
```

Comment:
This is a good start, I'm glad we have a test that shows how the feature is used! I think it would be even more valuable as a stress test, given that this test is looking for potential race conditions, and not really testing the feature's intended behavior. How about running ~100 writes?

Reply:
Probably? If an implementation offloads a callback when it shouldn't, that's when we run the risk of a use-after-free right? That would be a race that depends on which thread gets scheduled first. Ratcheting up the iteration count would do a better job of catching that.

Reply:
done